dag_in_context/src/type_analysis.egg

```diff
@@ -206,148 +218,70 @@
       :ruleset error-checking)
 
 ; Binary ops
-(rule (
-        (= lhs (Bop (Add) e1 e2))
-        (HasType e1 (Base (IntT)))
-        (HasType e2 (Base (IntT)))
-      )
-      ((HasType lhs (Base (IntT))))
-      :ruleset type-analysis)
-(rule ((= lhs (Bop (Add) e1 e2)))
-      (
-        (ExpectType e1 (Base (IntT)) "(Add) expected int, received {?}")
-        (ExpectType e2 (Base (IntT)) "(Add) expected int, received {?}")
-      )
-      :ruleset type-analysis)  
 
-(rule (
-        (= lhs (Bop (Sub) e1 e2))
-        (HasType e1 (Base (IntT)))
-        (HasType e2 (Base (IntT)))
-      )
-      ((HasType lhs (Base (IntT))))
-      :ruleset type-analysis)
-(rule ((= lhs (Bop (Sub) e1 e2)))
-      (
-        (ExpectType e1 (Base (IntT)) "(Sub) expected int, received {?}")
-        (ExpectType e2 (Base (IntT)) "(Sub) expected int, received {?}")
-      )
-      :ruleset type-analysis)  
-
-(rule (
-        (= lhs (Bop (Mul) e1 e2))
-        (HasType e1 (Base (IntT)))
-        (HasType e2 (Base (IntT)))
+;; Operators that have type Type -> Type -> Type
+;; Note we only do this generic matching for binary
+;; operator since there's a lot of them.
+;; In the future we can also extend to other constructs.
+(relation bop-of-type (BinaryOp Type))
+(bop-of-type (Add) (Base (IntT)))
+(bop-of-type (Sub) (Base (IntT)))
+(bop-of-type (Div) (Base (IntT)))
```

Comment:
Note that the old type checker does not do type checking for integer division, and after supporting this, pure integer benchmarks like `benchmarks/passing/bril/core/digital-root.bril` blow up.